### ISSUE TITLE
Tree lod data modification

### ISSUE THREAD

**Issue by n0p0n3** [TES5Edit/TES5Edit#868]
is there any way to change LOD mesh to an arbitrary mesh using xedit? I want to change the tree lod data to the same mesh as the regular ingame mesh to have high quality lod data to extract for my vr project. I'm trying to create a plugin for each type of object to extract their position, rotation, and scaling as they are in game. My imagined process is to create a xedit filter to block all other object types except the ones i'm extracting, change the objects record type to static (or just set a lod mesh as the same mesh as the player sees up close), then save it as a plugin so that anyone can then run xlodgen, thus having a combined mesh with all objects in their correct places and scaled. Then I can export the .btr through nifscope, load each cell in blender, separate each object into a separate mesh, then load that mesh in unity and assign scripts as needed.

I've already extracted the terrain data at 2k resolution, and can physically replace the lod meshes with ones extracted with bae for object lod, but figured just making a plugin would make the conversion pipeline easier to replicate (since i couldn't distribute due to licensing). Just distribute an .esp and instructions on how to extract.

tldr; I can't seem to find how xedit displays lod data or how to edit them

**Comment by zilav** [comment]
xLODGen is hardcoded to use CK rules for LOD assets assignment,
Use DynDOLOD which supports custom rules.

**Comment by n0p0n3** [comment]
so that will work for skyrim se, but for fallout 4 i'll have to use the creation kit to change the lod meshes myself? and 76, if they ever release the creation kit for it. Is there anyway to use DynDOLOD with 4/76?

**Comment by zilav** [comment]
Yes, LOD meshes are defined in the MNAM subrecord of statics, you can change them in xEdit too and use xLODGen.
FO76 doesn't support modding hence no tools.

**Comment by n0p0n3** [comment]
can xedit create an esp for fo76, even if it couldnt be used? im not actually using the games themselves, just the tools to "rip" the assets in the correct places and orientations.  If 76edit can create a modified esp and load it itself to overwrite the records so that I can then use xlodgen, that would be enough. Also, is there a way to convert objects to static objects? All I need is the xyz position, the xyz rotation, the scale, and the lod mesh field.
Thank you for the quick replies, and the awesome tools

**Comment by n0p0n3** [comment]
also, how would i alter the MNAM subrecord of statics? It's greyed out wherever I look at static objects. Basically, I just want to set MNAM to match MODL

edit: I figured out how to set the MNAM/LOD/Mesh field, but idk if I need to put anything in the LOD field or just the mesh field.
![Untitled](https://user-images.githubusercontent.com/54335729/94238775-76559c80-fed6-11ea-8094-5959e2e40590.png)

**Comment by zilav** [comment]
You specify LOD mesh nif file to use for the correrponding LOD level (4 entries, LOD levels 0..3, the first one is level 0 which is the closest).

**Comment by n0p0n3** [comment]
Ok, thank you. One last question then I'll stop taking up your valuable time lol. Is there a way to convert objects between record types, ie convert weapons/armor/consumables/ammo to statics so they can then be used to create a lodmesh like buildings? Or any way to get the MNAM field in any record. And less relevent at the moment (as I dont own it yet, i'm working on fallout 4 and skyrim, then the remake mods when they come out) is the question of 76edit being able to create esps, not to be loaded in the game, but to be used in 76edit itself to do the same as I'm doing with the others.
Thank you very much for your patience, all the guides I see about xedit are for cleaning mods, not what else it can do.

**Comment by sheson** [comment]
It is possible to instruct LODGen.exe to not group objects into a single shape/shader with DontGroup=True in the options/export file.

If I understand this right you want to create object LOD for as many things as possible to import for another engine?

You might just write your own script to create the export file for LODGen.exe, then you do no need to change any records/plugins.

The export file is a text file xLODGen/DynDOLOD creates with instruction for LODGen.exe how to build object LOD *BTO meshes. 

You can find it in [xEdit/xLODGen]\Edit Scripts\LODGen.txt or DynDOLOD\Edit Scripts\Export\LODGen_[GameMode]_[Worldspace].txt after generating object LOD.

After a header part you will find each object has tab separated line.
Refererence FormID, Referene Flags, Pos x, Pos y, Pos z, Rot x, Rot y, Rot z, Scale, EditorID, Base Record Flags, Material Shader text, full model, LOD Level 4 model, LOD Level 8 model, LOD Level 16 model, LOD Level 32 model
In case of FO4 there will be more data for SCOLs (combined objects).

You can run LODGen.exe yourself with your own export file to do most of what you want. Do not define any TextureAtlasMap in the header, so it leaves the texture paths alone for each model.

LODGenx64.exe --inputfile "C:\Path\To\File.txt"

**Comment by n0p0n3** [comment]
That sounds like you just laid everything out for me on a silver platter. Yes, I'm just trying to extract everything where it needs to be to be to be imported into another engine as is, where I'll hand write the new interactions. I'll have to go find guides on how to write an xedit script tomorrow (and a primer on pascal, I believe is used?). Thank you for the help and great tools.

**Comment by n0p0n3** [comment]
Is there any decent resources on learning xedit scripting from scratch? All I've seen are guides and resources that seem to expect a high level understanding to begin with.  I get that I need to create the export file and feed it to lodgen using the cmd line, it's the structure of the contents of the export file that I'm having difficulty with. I know I need the xyz pos, xyz rotation, scale, full model, lod level 4 model and the equivalents from combined objects. I only need the lod level 4 mesh, as I'm thinking of trying a new idea I had for lod/distance meshes.

**Comment by n0p0n3** [comment]
Am I correct in the assumption that I could take the lodgen script, copy it and change it to use the model mesh field instead of lod meshes, then use the filters to hide object types that I'm not trying to extract at the moment? From reading the lodgen script, 
Initilize is the main function, sets slodbaseobjects to stat+scol (for fallout), calls dolodgen(),  which calls exportworldspace(), which calls iterateworldspace(), which calls processreference(),  which calls lodmeshfor(), which determines which lod level to use. So if I force mFull to ilodFull (1), then it should force lodmeshfor() to export whatever I pass through the filter?

**Comment by n0p0n3** [comment]
It took awhile to figure some of this out, and I was initially just looking for a way to quickly fix the trees having different lod meshes than the model mesh, but your answer could solve my later problems with ripping weapons, armor, consumables, and ammo, so thank you for that. Until I can figure out how to write the script to create the export file, I can at least just change all the lod level 4 mesh paths to the full model meshes, but how would I create the export file? I think i've found how it's created in the processrefence function, would I just replace the m4 lod level 4 mesh path with mFull mesh path? Would I just iterate through every object of [type] in every cell for weapons, armor, consumables, and ammo?

**Comment by n0p0n3** [comment]
It took awhile to figure some of this out, and I was initially just looking for a way to quickly fix the trees having different lod meshes than the model mesh, but your answer could solve my later problems with ripping weapons, armor, consumables, and ammo, so thank you for that. I took the lodgen.pas script and edited it to remove the TextureAtlasMap stuff and changed the m4 variable that pointed to the level 4 lod mesh and replaced it with another mFull to use the full mesh instead, as well as adding the DontGroup variable. All thats left is to add the object base records for objects that are physically placed in the world to sLODBaseObjects. Could one of you look it over when you can and critique it please? I'm also not able to select the worldspace.
[asset ripper.txt](https://github.com/TES5Edit/TES5Edit/files/5287726/asset.ripper.txt)

**Comment by sheson** [comment]
The script exports data based on the "mesh rules".
The last rule with "/" basically matches all full models that were found for the signature in sLODBaseObjects.
The VWD only flag ensures that only data for objects is exported that have base records with HasDistantLOD (VisibleWhenDistant) flag set.
For your case you want to export everything, so you want to have two rules that are the same one with VWD checked and one with it unchecked so it also export all reference with base records that do not have the flag set.
Or you just remove all the rule  stuff in the code and the VWD check and simply export everything.

The slLODMeshes seems to be something left over not used anymore.

**Comment by n0p0n3** [comment]
I can't seem to test it with fallout 4, I can't select a worldspace. It gave me the only supports f3, nv, and skyrim error until I added wbGameMode = gmFO4, but I still cannot select a worldspace to export to test on, and I can't hook up my vr headset for awhile to install drivers, so I can't test skyrimvr. any ideas?

**Comment by n0p0n3** [comment]
So, If I'm correct, then these seem to be the relevent sections?

// VWD only?
      if j and $01000000 = $01000000 then
        if not GetIsVisibleWhenDistant(stat) then
          Continue;
      if Pos(slLOD[i], mFull) > 0 then begin
        m4 := mFull;
        // other LOD levels only for Skyrim
        if wbGameMode = gmTES5 then begin
          m8 := mFull;
          m16 := mFull;
        end;

 // a tab separated string of Editor ID, flags, material, full mesh and lod files
          s := EditorID(stat) + #9 + IntToHex(GetElementNativeValues(stat, 'Record Header\Record Flags'), 8) + #9 +
               mat + #9 + mFull + #9 + mFull + #9 + mFull + #9 + mFull;
          idx := slCache.Count;
          slCache.AddObject(s, statfid);
          
 // Fallout: High Priority LOD info variation with m4 model for m8, at the same index as normal cache
          if bFallout then begin
            s := EditorID(stat) + #9 + IntToHex(GetElementNativeValues(stat, 'Record Header\Record Flags'), 8) + #9 +
                 mat + #9 + mFull + #9 + mFull + #9 + mFull + #9 + mFull;
            slCacheHPLod.Add(s)
          end;

 // not a lod object: no matching lod rules and not in cache
  if idx = -1 then begin
    lstSkip.Add(statfid);
    Exit;
  end;

so by removing the 
// VWD only?
      if j and $01000000 = $01000000 then
        if not GetIsVisibleWhenDistant(stat) then
          Continue; 

loop and the 
 
// not a lod object: no matching lod rules and not in cache
  if idx = -1 then begin
    lstSkip.Add(statfid);
    Exit;

loop, and why ever I can't select a worldspace, it should work? I'm sorry to keep bothering yall, but I can't begin the rest of the work until I have terrain and the statics because I'm going to use a megamesh and megatexture style system to test performance since I'm aiming for the oculus quest 2, and I need the objects in their correct positions to begin applying scripts to replicate game effects

**Comment by n0p0n3** [comment]
I want for statics, static collections, and anything that can't be interacted with or doesn't need physics applied to it to be fused into one megamesh and megatexture per cell. Anything that can be interacted with or does need physics applied I want extracted as one mesh, but each object as a separate manipulable mesh not joined. So statics, static collections, and flora I guess (the plant bodies at least) will be one fused mesh and texture. Everything else will be handled like normal, and I'll do lod stuff either as a set of sprite billboards with baked on textures with a raytracer to determine which billboard to use, or photogrametry generated using the cell meshes. I'll basically need to have the full environmental setup complete in order to create the lod stuff because I want to have the furthest cells in view distance to represent all the cells beyond them

**Comment by sheson** [comment]
You replaced a lot of code that made the rules obsolete. E.g. instead of changing code everywhere to use retrieve and use the full model, you could have just changed the mesh rules to "Full Model". So, yes, just remove the VWD check in the code and everything related to rules, so it should export everything. I would only set LOD Level 4 in the export file, or instruct LODGen to run with --LODLevel 4 so you do not waste time generating the files for LOD levels not required.

Best not to treat FO4 like the old Fallouts. It is more similar to Skyrim (gmTES5)  / Skyrim Special Edition (gmSSE). In fact, I have an old [LODGenFO4.zip](https://github.com/TES5Edit/TES5Edit/files/5291726/LODGenFO4.zip) here from before the code was added in xEdit itself. It should help with exporting FO4 SCOLs. Also [check the implementation in wbLOD.pas](https://github.com/TES5Edit/TES5Edit/blob/dev-4.1.4/Core/wbLOD.pas). 

You need to figure out yourself the best way to filter the export for your needs. Base record signatures of course is already implemented, then maybe other flags than HasDistantLOD, skip IsMarker flag for example.

Really, this is just a start that can go into a really deep hole of details and special cases. Since LOD only supports a subset of shaders and settings, add PassThruMeshMask=nif to the export header, to allow for more shader settings to be untouched, but it will still set/unset shaders and settings that LOD does not support. If that starts to hinder things, let me know.

**Comment by n0p0n3** [comment]
I ran the lodgenfo4.pas and I can finally select a worldspace. I'm going to modify it to do what I want. Since the rules will allow me to select the full model, I'm going to leave them alone (don't want to break anything). I'm going to remove the vwd check (does unchecking it in the form disable it?), the build atlas stuff (does unchecking it disable it too?) and add all the base record types to sLODBaseObjects, and probably just use the filter by base record to pick and choose what I want to rip at that moment. I didn't get much done today or last night because I had a broken tooth and had to get it pulled this morning, and I'm still kind of woozy, so it may be a few days before I can do any real testing or editing, but thank you for all your help. Also, just curious, what is wrong with the skyrim meshes that are in sMeshIgnoreTranslationTES5?

edit: I uploaded the modified fo4lodgen.pas script, i pretty much just commented out the vwd section, added in the new sLODBaseObjects, and left a note to remind me to add a check for MODL field, because not every object has a MODL and I didn't know if it would crash or just skip it.
[Fo4 Asset ripper.txt](https://github.com/TES5Edit/TES5Edit/files/5296392/Fo4.Asset.ripper.txt)

**Comment by sheson** [comment]
if I mentioned earlier if you only use rules, then two rules need to be added, one with VWD check and one unchecked. Removing the check from the script works as well.

If you uncheck build atlas, no atlas texture will be created. No atlas is needed in your case, obviously.

sLODBaseObjects is a list of vanilla LOD meshes that have translation/rotation on that must be ignored.

There should be no need to specifically check for no model. The return result will be empty and that should already be handled.

**Comment by n0p0n3** [comment]
So I'm running a test right now because I've discovered that I'm an idiot. I had the filter by base references radio mark ticked, but nothing selected under references. That's fixed, and why i haven't been getting any results at all. Now I'm getting an export file that seems to have the meshes listed correctly, but I'm getting an atlas output in /data/textures/terrain/commonwealth/objects even with the atlas deselected, and no files in data/meshes/terrain/commonwealth/objects, even when I run lodgen from the cmd line with the export files as input.

It's so close I can almost taste it. I hope it's just a silly thing I haven't noticed and you'll be able to point out to me like I'm  a child.

[FO4 LODgen Commonwealth.txt](https://github.com/TES5Edit/TES5Edit/files/5309564/FO4.LODgen.Commonwealth.txt)
I can't upload the full export file because it's 256Mb in size!

**Comment by sheson** [comment]
The script is hard coded to not generate the atlas? Is it actually spending time on generating textures?

Get the latest LODGen.exe/LODGenx64.exe from the current xLODGen beta 64. That should generate some output with the input file.

Actually, the header setting to keep shapes separated is DontMergeShapes=True. I remembered wrong.

**Comment by n0p0n3** [comment]
One of the first sections is 
// Atlas options
  bBuildAtlas = True;
  iAtlasTextureSize = 512; // max size of source lod texture to be included on atlas
  iAtlasTileSize = 512; // max size of tile on atlas, source texture is rescaled if above
  iAtlasWidth = 8192;
  iAtlasHeight = 8192;
  fUVRange = 1.5; // put textures on atlas having -UVRange <= uv <= UVRange in lod nif models

and later in exportworldspace section

// use pre-defined vanilla atlas for now
  if bFO4 and not bBuildAtlas then
    AtlasMapName := ScriptsPath + 'FO4' + EditorID(wrld) + 'LODGenAtlasMap.txt';

 then later in the main section

if (wbGameMode = gmFO4) then begin
    bFO4 := True;
    bBuildAtlas := False;
  end
  else
    bFO4 := False;

I've already changed it to be False in the beginning, and I've run it in 
xLODGen 4.1.3b EXTREMELY EXPERIMENTAL Terrain LOD BETA (50FACBFD) and xedit 4.0.3. Neither will output to the terrain/commonwealth/objects folder.

**Comment by sheson** [comment]
```
if (wbGameMode = gmFO4) then begin
bFO4 := True;
bBuildAtlas := False;
```
in the init section means it should be hardcoded to not execute the atlas generation code that starts with 
`if bBuildAtlas then begin ...`


The truncated file you uploaded generates output with LODGen.exe 2.7.8.0 from latest xLODGen beta when executed with
lodgenx64.exe --inputfile "FO4.LODgen.Commonwealth.txt" --lodlevel 4

Obviously I updated the paths D:\SteamLibrary to match my setup. Check the log messages for hints.

**Comment by n0p0n3** [comment]
I'm getting output now. I guess it was a combination of the difference between xlodgen (which I was running) and xlodgen 64 (I didn't run because it gave a message saying that it took extra resources and to use the 32 bit version if you didn't need to use the 64bit version, and I just never tried it again after that), and that I forgot to add --lodlevel 4 when I ran it. I'm not getting any textures when I check the output in nifscope, but thats probably because I haven't used bae to extract the resources yet. Thank you for all your assistance, I seriously couldn't have figured this out without a guiding hand. Now I can begin the more "fun" parts of the conversions instead of the tedious hand placing of objects. Ya'll just saved me 1000's of hours of work, maybe even years. I'll have to ask for help again once I get to interiors, but I don't foresee that being until after I get the overworld for skyrim and fallout 4 done. I intend to release the parts that I'm writing and a how to guide on how to do what ya'll have helped me find out, that way anyone can take their existing installation and convert it similar to the daggerfallunity project. Once i get a stable build going, I'll find a way to contact ya'll for a preview release.

**Comment by n0p0n3** [comment]
I may actually begin with skygerfall, as it's much simpler than base skyrim or fallout 4, but I'm going to do all the "dlc sized remasters" such as skywind, skyblivion, F4: capital wasteland, F4NV, Enderal: forgotten stories, Fallout 76, and maybe even FNV: Dust, and F4: Frost. I'm aiming for the oculus quest 2, but I currently own a Windows Mixed Reality headset, so I'll be supporting steamvr as well.

**Comment by n0p0n3** [comment]
What does the "group=false/true" due in xlodgen64 and what is the variable associated with it?

**Comment by sheson** [comment]
Nothing useful for LOD generation. It used to prevent translation/merging of shapes for re-UVing meshes. It requires tab #9 - which contains the flags from the base record - to have bit 1 set to have any effect.

**Comment by n0p0n3** [comment]
Ok thanks. Is there any way to export lodgen for interior cells? Other than copying the interior and pasting it in a new worldspace. Just wondering while I redo what I had done because I didn't realize the script was exporting everything even though I had filters on, and I have to wait again.

**Comment by sheson** [comment]
Call wbFindREFRsByBase(e, sLODBaseObjects, f, lst) with e being the internal cell record to find all its children references.

**Comment by n0p0n3** [comment]
How would I get interior cell records? Can I alter the script to add interiors to the worldspace list? Where is e even defined in the script? Also, is there a way for the terrain lodgen script to output just the water meshes? The terrain is fine (except maybe generating it as a solid cube - the terrain top), but I'm having to go through and export the water separately from the land, and if I can export just the water on a separate run, that will speed up alot of processing.

**Comment by n0p0n3** [comment]
Overall, the changes I'm trying to make are have wbFindREFRsByBase in a loop, with each call's sLODBaseObjects being set as the element of an array that gets iterated at the end of the loop (I get all subrecord types exported even when I have a filter set for specific types), have interiors iterated through and exported at the end of each worldspaces turn (worldspace 1 runs throught all exterior cells exporting one type of subrecord, then begins running through all interior cells exporting the same subrecord, then loops through again for the next subrecord, etc). And exporting water lod by itself during terrain lod, possibly as a box shape of depth equal to the land depth instead of as a plane? I'm going to do a subtraction modifier in blender to make the water as a solid mesh fitting the land under the surface so I can have an  on-collision script for swimming animations and movement.

**Comment by sheson** [comment]
Without actually programming it myself, I am not a 100% sure. Maybe Zilav can help. Getting all 'CELL' main records instead of the 'WRLD' main records in a copy of FillWorldspacesWithLOD might work, then use wbGetSiblingRecords on it. Else the older way of looping through the block and subblocks like in "Find cell in worldspace.pas".

The only available option is to not generate any water LOD, since there is a No LOD Water DATA flag that can be set on a worldspace record. The other way around is not implemented.

**Comment by n0p0n3** [comment]
I've tried replacing WRLD with CELL in several places in FillWorldSpaceWithLOD function, all I get is "access violation at address 0040A400 in module 'fo4edit.exe'. Write of address 00000004." error. How would I get ahold of zilav to ask about it? He was helping me, but I don't know how to ask except for starting another issue, and I don't want to do that if unnecessary to avoid clogging the bugfix line. I'm looking in "Find cell in worldspace.pas", but until I have time to walk through the whole script thoroughly , I don't understand what to look at or how to use it yet. The water thing is too bad, but not a problem, I'd already processed almost half the commonwealth level 4 terrain meshes before I even thought to ask about it. I just have to do some manual processing on each cell to make the water into a block, and make the terrain mesh manifold before I can do a boolean operation to create a mesh "casting" of the water that I can use for detecting if player is in water to switch to swimming animations and other things like breath holding.

**Comment by zilav** [comment]
xEdit has it's own discord server for such discussions, linked to one of the buttons in the right upper corner.
